tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	}
+	lastSeq := seqs[0]
+	for _, seq := range seqs[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential head chunk files %d and %d", lastSeq, seq)
+		}
+		lastSeq = seq
+	}
+
+	for i, b := range cdm.mmappedChunkFiles {
+		if b.Len() < HeadChunkFileHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid head chunk file header in file %d", i)
+		}
+		// Verify magic number.
+		if m := binary.BigEndian.Uint32(b.Range(0, MagicChunksSize)); m != MagicHeadChunks {
+			return errors.Errorf("invalid magic number %x", m)
+		}
+
+		// Verify chunk format version.
+		if v := int(b.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+8))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last head chunk file. As a safe buffer, we set the maxt to mint + 1.5 times head chunk file time range.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer func() {
+				var merr tsdb_errors.MultiError
+				merr.Add(returnErr)
+				merr.Add(f.Close())
+				returnErr = merr.Err()
+			}()
+			mint := binary.BigEndian.Uint64(b.Range(HeaderMintOffset, HeaderMintOffset+8))
+			binary.BigEndian.PutUint64(cdm.writeBuf[:], mint+(uint64(cdm.maxFileMs)*3/2))
+			if _, err := f.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		cdm.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func sequenceFilesMap(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (cdm *ChunkDiskMapper) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chunkRef uint64, err error) {
+	cdm.wbufLock.Lock()
+	defer cdm.wbufLock.Unlock()
+
+	if cdm.closed {
+		return 0, ErrChunkDiskMapperClosed
+	}
+
+	if cdm.shouldCutNewFile(len(chk.Bytes()), maxt) {
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+		cdm.curFileMaxt = 0
+	}
+
+	// seriesRef(8) + mint(8) + maxt(8) + encoding(1) + max uvarint size (5) + the bytes is
+	// the total size to be written.
+	// if len(chk.Bytes()) >= writeBufferSize-30, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes()) < writeBufferSize-30 && cdm.wbuf.Available() < 8+8+8+1+5+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
```

Comment:
Can we always just flush?

Reply:
That was what I had initially, flush at the end of WriteChunk. But to avoid write amplification, we are buffering it.

Reply:
hm that is quite interesteding, have you done any research? could there be some way to actually test the difference in the write amplification?

Reply:
> Isn't the kernel here keeping some write buffer or is this synced at a set interval

It does keep some buffer but I am not sure how it interacts with flush here. I _guess_ it has to sync with the disk for it to me m-mapped.

And regarding how much amplification we can expect, here is some quick calculation:

SSD page size starts at 2KB (from a quick Google search). And the chunks being written to the disk are under 200-250B (including the meta). Let's consider it to be 256B for easy calculation. So in that case, if we had to flush chunk for every write, we would be doing >8x more disk writes in terms of volume. And this is just for an empty page.

Given a single page can accommodate multiple chunks, overwriting the same page is a bigger task which includes copying+erasing+writing the new updated content at _block_ level. SSD block sizes vary between 256KB-4MB (from a quick Google search). So, in this case, it turns out to be 1204-16384x more writes by volume.

Reply:
I think that is one of the reasons we have the "page" concept in the WAL too. And have 8MB write buffer when writing a block. Now that makes me feel I should increase the write buffer size to 4MB here from 2MB that I have right now.